hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/DataSkippingUtils.scala

```diff
@@ -270,35 +344,63 @@ object DataSkippingUtils extends Logging {
             val targetExprBuilder: Expression => Expression = swapAttributeRefInExpr(sourceExpr, attrRef, _)
             val minValueExpr = targetExprBuilder.apply(genColMinValueExpr(colName))
             val maxValueExpr = targetExprBuilder.apply(genColMaxValueExpr(colName))
             Not(And(StartsWith(minValueExpr, value), StartsWith(maxValueExpr, value)))
-          }
+          }.orElse({
+          hasNonIndexedCols.set(true)
+          Option.empty
+        })
 
       case or: Or =>
-        val resLeft = createColumnStatsIndexFilterExprInternal(or.left, indexSchema)
-        val resRight = createColumnStatsIndexFilterExprInternal(or.right, indexSchema)
-
-        Option(Or(resLeft, resRight))
+        val leftHasNonIndexedCols = new AtomicBoolean(false)
+        val resLeft = createColumnStatsIndexFilterExprInternal(or.left, isExpressionIndex = isExpressionIndex,
+          indexedCols = indexedCols, hasNonIndexedCols = leftHasNonIndexedCols)
+        val rightHasNonIndexedCols = new AtomicBoolean(false)
+        val resRight = createColumnStatsIndexFilterExprInternal(or.right, isExpressionIndex = isExpressionIndex,
+          indexedCols = indexedCols, hasNonIndexedCols = rightHasNonIndexedCols)
+        if (leftHasNonIndexedCols.get() || rightHasNonIndexedCols.get()) {
+          hasNonIndexedCols.set(true)
+          None
+        } else {
+          Option(Or(resLeft, resRight))
+        }
 
       case and: And =>
-        val resLeft = createColumnStatsIndexFilterExprInternal(and.left, indexSchema)
-        val resRight = createColumnStatsIndexFilterExprInternal(and.right, indexSchema)
-
-        Option(And(resLeft, resRight))
+        val leftHasNonIndexedCols = new AtomicBoolean(false)
+        val resLeft = createColumnStatsIndexFilterExprInternal(and.left, isExpressionIndex = isExpressionIndex,
+          indexedCols = indexedCols, hasNonIndexedCols = leftHasNonIndexedCols)
+        val rightHasNonIndexedCols = new AtomicBoolean(false)
+        val resRight = createColumnStatsIndexFilterExprInternal(and.right, isExpressionIndex = isExpressionIndex,
+          indexedCols = indexedCols, hasNonIndexedCols = rightHasNonIndexedCols)
+        // only if both left and right has non indexed cols, we can set hasNonIndexedCols to true.
+        // If not, we can still afford to prune files based on col stats lookup.
+        if (leftHasNonIndexedCols.get() && !rightHasNonIndexedCols.get()) {
+          Option(resRight)  // if only left has non indexed cols, ignore from the expression to be looked up in col stats df
+        } else if (!leftHasNonIndexedCols.get() && rightHasNonIndexedCols.get()) {
+          Option(resLeft) // if only right has non indexed cols, ignore from the expression to be looked up in col stats df
+        } else if (leftHasNonIndexedCols.get() && rightHasNonIndexedCols.get()) {
+          hasNonIndexedCols.set(true)
+          None
+        } else {
+          Option(And(resLeft, resRight))
+        }
```

Comment:
Can be slightly simplified (less atomic boolean checks):
```
        if (leftHasNonIndexedCols.get() && rightHasNonIndexedCols.get()) {
          hasNonIndexedCols.set(true)
          Option.empty
        } else if (leftHasNonIndexedCols.get()) {
          Option(resRight) // Ignore the non-indexed left expression
        } else if (rightHasNonIndexedCols.get()) {
          Option(resLeft) // Ignore the non-indexed right expression
        } else {
          Option(And(resLeft, resRight))
        }
```

Reply:
sg.